### DISCUSSION TITLE
ability to add plugins like wordpress

### DISCUSSION THREAD

**Discussion by sarmenb** [laravel/framework#39512] [Ideas]
it would be great if there was a way to add plugins into a laravel application like how wordpress plugins are installed without having to modify composer.

**Comment by dennisprudlo** [comment]
Well you don't need to modify composer. The thing is that you need to get the "plugin" from somewhere. Composer helps to download it into your app in a standardized way. But you could also just put your plugin files in a separate folder in your app and use them. But here's the thing:

To load "plugins" like in wp is more a feature for an application not for a framework. You could use Laravel to build your own CMS and implement a feature that users can install plugins.

Is there a particular reason why you wouldn't want to modify the composer file?

**Comment by ITwrx** [comment]
> To load "plugins" like in wp is more a feature for an application not for a framework. You could use Laravel to build your own CMS and implement a feature that users can install plugins.

This is the key. Laravel is a framework, not a CMS. Wordpress is both, but you are describing a feature of the CMS part. 

Here's a couple Laravel-based CMSes that appear to have modules/plugins: 

https://www.pyrocms.com/
https://octobercms.com/

**Reply by dennisprudlo** [reply]
Might add https://statamic.com/ to that list 😉

**Reply by ITwrx** [reply]
cool, thanks. 
yeah, i don't know about all the Laravel CMSes. They are soooo NIH! lol. i had just done a quick search and saw those two. :)

**Comment by sarmenb** [comment]
how would i implement this feature myself in my own app? anyone know of a tutorial? or i have to dig through these cmses and figure it out.

Is it kind of like installing a composer package and exporting assets to their locations on the fly? so if i setup in my gui to run a artisan command that executes composer install app_name then php artisan vendor:publish #3  or something.

**Reply by dennisprudlo** [reply]
This highly depends on what a plugin is for you. Is it like you have different functionality and users can "install"/activate those features. Should the user have the ability to create its own plugins? Then what should be able to do with those plugins? Is you app a SaaS oder something the users hosts on his own server. If SaaS, do we really want to allow users to install code on our server?

> Is it kind of like installing a composer package and exporting assets to their locations on the fly? so if i setup in my gui to run a artisan command that executes composer install app_name then php artisan vendor:publish #3 or something.
I wouldn't recommend this. Well if the user is also the owner of the server and we assume he knows what he does then its fine. I would install plugins somewhere in the storage directory so its not actually part of my application.

Do you have a sample use case for a user who needs to install a specific plugin in your app? This would help to understand whats the actual goal here.
